File: apps/mobile/src/services/notifications.ts
import type { SupabaseClient } from "@supabase/supabase-js";
import { authHeader, getApiBaseUrl } from "./api-client";

// ── Types ────────────────────────────────────────────────────────────────

export type NotificationPreferences = {
  enabled: boolean;
  daily_reminder_enabled: boolean;
  daily_reminder_time: string;
  budget_alert_enabled: boolean;
  budget_alert_thresholds: number[];
  weekly_summary_enabled: boolean;
  weekly_summary_day: number;
  weekly_summary_time: string;
  ai_insight_enabled: boolean;
  bill_reminder_enabled: boolean;
  timezone: string;
  push_enabled: boolean;
};

export type NotificationItem = {
  id: string;
  type: string;
  title: string;
  body: string;
  data: Record<string, unknown>;
  read_at: string | null;
  created_at: string;
};

export type NotificationListResponse = {
  items: NotificationItem[];
  unread_count: number;
};

// ── Preferences ──────────────────────────────────────────────────────────

export async function getNotificationPreferences(
  supabase: SupabaseClient
): Promise<NotificationPreferences> {
  const res = await fetch(
    `${getApiBaseUrl()}/api/v1/notifications/preferences`,
    { headers: { ...(await authHeader(supabase)) } }
  );
  if (!res.ok)
    throw new Error(`Gagal memuat preferensi notifikasi (${res.status})`);
  return res.json();
}

export async function updateNotificationPreferences(
  supabase: SupabaseClient,
  patch: Partial<NotificationPreferences>
): Promise<NotificationPreferences> {
  const res = await fetch(
    `${getApiBaseUrl()}/api/v1/notifications/preferences`,
    {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
        ...(await authHeader(supabase)),
      },
      body: JSON.stringify(patch),
    }
  );
  if (!res.ok)
    throw new Error(`Gagal menyimpan preferensi notifikasi (${res.status})`);
  return res.json();
}

// ── Notification list ────────────────────────────────────────────────────

export async function listNotifications(
  supabase: SupabaseClient,
  options?: { limit?: number; unreadOnly?: boolean }
): Promise<NotificationListResponse> {
  const params = new URLSearchParams();
  if (options?.limit) params.set("limit", String(options.limit));
  if (options?.unreadOnly) params.set("unreadOnly", "true");
  const qs = params.toString();
  const url = `${getApiBaseUrl()}/api/v1/notifications${qs ? `?${qs}` : ""}`;

  const res = await fetch(url, {
    headers: { ...(await authHeader(supabase)) },
  });
  if (!res.ok)
    throw new Error(`Gagal memuat notifikasi (${res.status})`);
  return res.json();
}

// ── Unread count ─────────────────────────────────────────────────────────

export async function getUnreadNotificationCount(
  supabase: SupabaseClient
): Promise<number> {
  const res = await fetch(
    `${getApiBaseUrl()}/api/v1/notifications/unread-count`,
    { headers: { ...(await authHeader(supabase)) } }
  );
  if (!res.ok)
    throw new Error(`Gagal memuat jumlah notifikasi (${res.status})`);
  const data = await res.json();
  return data.unread_count ?? 0;
}

// ── Mark read ────────────────────────────────────────────────────────────

export async function markNotificationRead(
  supabase: SupabaseClient,
  id: string
): Promise<void> {
  const res = await fetch(
    `${getApiBaseUrl()}/api/v1/notifications/${id}/read`,
    {
      method: "PATCH",
      headers: { ...(await authHeader(supabase)) },
    }
  );
  if (!res.ok)
    throw new Error(`Gagal menandai notifikasi terbaca (${res.status})`);
}

export async function markAllNotificationsRead(
  supabase: SupabaseClient
): Promise<void> {
  const res = await fetch(
    `${getApiBaseUrl()}/api/v1/notifications/read-all`,
    {
      method: "PATCH",
      headers: { ...(await authHeader(supabase)) },
    }
  );
  if (!res.ok)
    throw new Error(`Gagal menandai semua terbaca (${res.status})`);
}

File: backend/app/models/notifications.py
"""Pydantic models for notification preferences and notification items."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class NotificationPreferences(BaseModel):
    """Full notification preferences for a user."""

    enabled: bool = True
    daily_reminder_enabled: bool = True
    daily_reminder_time: str = Field(
        default="20:00", pattern=r"^([01][0-9]|2[0-3]):[0-5][0-9]$"
    )
    budget_alert_enabled: bool = True
    budget_alert_thresholds: list[int] = [80, 100]
    weekly_summary_enabled: bool = True
    weekly_summary_day: int = Field(default=0, ge=0, le=6)
    weekly_summary_time: str = Field(
        default="19:00", pattern=r"^([01][0-9]|2[0-3]):[0-5][0-9]$"
    )
    ai_insight_enabled: bool = True
    bill_reminder_enabled: bool = False
    timezone: str = "Asia/Jakarta"
    push_enabled: bool = False


class NotificationPreferencesUpdate(BaseModel):
    """Partial update for notification preferences."""

    enabled: bool | None = None
    daily_reminder_enabled: bool | None = None
    daily_reminder_time: str | None = Field(
        default=None, pattern=r"^([01][0-9]|2[0-3]):[0-5][0-9]$"
    )
    budget_alert_enabled: bool | None = None
    budget_alert_thresholds: list[int] | None = None
    weekly_summary_enabled: bool | None = None
    weekly_summary_day: int | None = Field(default=None, ge=0, le=6)
    weekly_summary_time: str | None = Field(
        default=None, pattern=r"^([01][0-9]|2[0-3]):[0-5][0-9]$"
    )
    ai_insight_enabled: bool | None = None
    bill_reminder_enabled: bool | None = None
    timezone: str | None = None
    push_enabled: bool | None = None


class NotificationItem(BaseModel):
    """A single notification record."""

    id: str
    type: str
    title: str
    body: str
    data: dict[str, Any] = Field(default_factory=dict)
    read_at: datetime | None = None
    created_at: datetime


class NotificationListResponse(BaseModel):
    """Paginated notification list with unread count."""

    items: list[NotificationItem]
    unread_count: int


class UnreadCountResponse(BaseModel):
    """Simple unread count response."""

    unread_count: int
